builtin/credential/aws/path_login.go

```diff
@@ -0,0 +1,492 @@
+		return nil, err
+	}
+
+	resp := &logical.Response{
+		Auth: &logical.Auth{
+			Policies: policies,
+			Metadata: map[string]string{
+				"instance_id":      identityDoc.InstanceID,
+				"role_tag_max_ttl": rTagMaxTTL.String(),
+			},
+			LeaseOptions: logical.LeaseOptions{
+				Renewable: true,
+				TTL:       b.System().DefaultLeaseTTL(),
+			},
+		},
+	}
+
+	// Enforce our image/role tag maximum TTL
+	if maxTTL < resp.Auth.TTL {
+		resp.Auth.TTL = maxTTL
+	}
+
+	return resp, nil
+
+}
+
+// fetchRoleTagValue creates an AWS EC2 client and queries the tags
+// attached to the instance identified by the given instanceID.
+func fetchRoleTagValue(s logical.Storage, tagKey string) (string, error) {
+	ec2Client, err := clientEC2(s)
+	if err != nil {
+		return "", err
+	}
+
+	// Retrieve the instance tag with a "key" filter matching tagKey.
+	tagsOutput, err := ec2Client.DescribeTags(&ec2.DescribeTagsInput{
+		Filters: []*ec2.Filter{
+			&ec2.Filter{
+				Name: aws.String("key"),
+				Values: []*string{
+					aws.String(tagKey),
+				},
+			},
+		},
+	})
+	if err != nil {
+		return "", err
+	}
+
+	if tagsOutput.Tags == nil ||
+		len(tagsOutput.Tags) != 1 ||
+		*tagsOutput.Tags[0].Key != tagKey ||
+		*tagsOutput.Tags[0].ResourceType != "instance" {
+		return "", nil
+	}
+
+	return *tagsOutput.Tags[0].Value, nil
+}
+
+// handleRoleTagLogin is used to fetch the role tag if the instance and verifies it to be correct.
+// Then the policies for the login request will be set off of the role tag, if certain creteria satisfies.
+func handleRoleTagLogin(s logical.Storage, identityDoc *identityDocument, imageEntry *awsImageEntry) (*roleTagLoginResponse, error) {
+
+	// Make a secondary call to the AWS instance to see if the desired tag is set.
+	// NOTE: If AWS adds the instance tags as meta-data in the instance identity
+	// document, then it is better to look this information there instead of making
+	// another API call. Currently, we don't have an option but make this call.
+	rTagValue, err := fetchRoleTagValue(s, imageEntry.RoleTag)
+	if err != nil {
+		return nil, err
+	}
+
+	if rTagValue == "" {
+		return nil, fmt.Errorf("missing tag with key %s on the instance", imageEntry.RoleTag)
+	}
+
+	// Check if the role tag is blacklisted.
+	blacklistEntry, err := blacklistRoleTagEntry(s, rTagValue)
+	if err != nil {
+		return nil, err
+	}
+	if blacklistEntry != nil {
+		return nil, fmt.Errorf("role tag is blacklisted")
+	}
+
+	rTag, err := parseRoleTagValue(rTagValue)
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure that the policies on the RoleTag is a subset of policies on the image
+	if !strutil.StrListSubset(imageEntry.Policies, rTag.Policies) {
+		return nil, fmt.Errorf("policies on the role tag must be subset of policies on the image")
+	}
+
+	// Create a HMAC of the plaintext value of role tag and compare it with the given value.
+	verified, err := verifyRoleTagValue(s, rTag)
```

Comment:
Minor "hygiene" issue, but I think it's generally considered good practice to verify a MAC ASAP in any processing code of yours, lest you do meaningful work on garbage data. The class of concerns is roughly around leaking information via timing and getting other code to misbehave with crafted bad input. The latter case is probably less concerning in a memory-safe language like Go, but it can't hurt, right?